ethcore/src/state_db.rs

```diff
@@ -222,11 +341,46 @@ impl StateDB {
 	/// or if the account is not cached.
 	pub fn get_cached<F, U>(&self, a: &Address, f: F) -> Option<U>
 		where F: FnOnce(Option<&mut Account>) -> U {
-		if !self.is_canon {
+		let mut cache = self.account_cache.lock();
+		if !Self::is_allowed(a, &self.parent_hash, &cache.modifications) {
 			return None;
 		}
-		let mut cache = self.account_cache.lock();
 		cache.accounts.get_mut(a).map(|c| f(c.as_mut()))
 	}
+
+	// Check if the account can be returned from cache by matching current block parent hash against canonical
+	// state and filtering out account modified in later blocks.
+	fn is_allowed(addr: &Address, parent_hash: &Option<H256>, modifications: &VecDeque<BlockChanges>) -> bool {
```

Comment:
isn't this a relatively expensive check to do on each cache lookup? Can it be made more efficient by creating a view over a specific subset of the cache (excluding non-canonical entries based on a given block hash) which can be queried in constant time?